chain/client/src/chunk_validation.rs

```diff
@@ -146,6 +147,62 @@ impl ChunkValidator {
         });
         Ok(())
     }
+
+    /// This function does two things
+    ///     1. Verifies if the chunk_endorsements have enough stake for chunk to be included in block.
+    ///     2. Collects the chunk_endorsement signatures from chunk_validators to include in block body.
+    /// We return an optional signature.
+    ///     None specifies we didn't have enough stake to include chunk in block.
+    ///     Some(signature) gives us the ordered chunk_endorsement signatures to include in block.
+    pub fn get_chunk_endorsement_signature(
+        &mut self,
+        chunk_header: &ShardChunkHeader,
+    ) -> Result<Option<ChunkEndorsementSignatures>, Error> {
```

Comment:
You can make the various stages of chunk validation explicit and type safe by introducing separate types for each stage of validation. Then you can set explicit expectations on how validated you want your chunk in the method signature. This way the compile, the immutability and the determinism of chunk headers will guarantee safety for you.